org.eclipse.xtext.tests/src/org/eclipse/xtext/formatting2/internal/JavaFormatterGrammarTest.java

```diff
@@ -0,0 +1,200 @@
+import com.google.inject.Injector;
+
+/**
+ * @author Arne Deutsch - Initial contribution and API
+ */
+@RunWith(XtextRunner.class)
+@InjectWith(XtextInjectorProvider.class)
+public class JavaFormatterGrammarTest extends AbstractXtextTests {
+
+	private static final String NL = System.lineSeparator();
+	private static final String TAB = "\t";
+
+	@Inject
+	private FormatterTestHelper formatterTestHelper;
+
+	@Override
+	public void setUp() throws Exception {
+		super.setUp();
+		with(XtextStandaloneSetup.class);
+	}
+
+	@Test
+	public void action() {
+		// @formatter:off
+		assertFormattedGrammar(
+				"Rule:" + NL +
+				TAB + "SubRule ({ Other . source = current } op='.');"
+				,
+				"Rule:" + NL +
+				TAB + "SubRule ({Other.source=current} op='.');"
+		);
+		// @formatter:on
+	}
+
+	@Test
+	public void enumLiterals() {
+		// @formatter:off
+		assertFormattedGrammar(
+				"enum StateType:" + NL +
+				TAB + "FIRST|SECOND|THIRD;"
+				,
+				"enum StateType:" + NL +
+				TAB + "FIRST | SECOND | THIRD;"
+				);
+		// @formatter:on
+	}
+
+	@Test
+	public void enumLiteralAssignment() {
+		// @formatter:off
+		assertFormattedGrammar(
+				"enum StateType:" + NL +
+				TAB + "PSEUDO = 'cond';"
+				,
+				"enum StateType:" + NL +
+				TAB + "PSEUDO='cond';"
+				);
+		// @formatter:on
+	}
+
+	@Test
+	public void hiddenClause() {
+		// @formatter:off
+		assertFormattedGrammar(
+				"Child  hidden ( WS ,  ID ):" + NL +
+				TAB + "A;"
+				,
+				"Child hidden(WS, ID):" + NL +
+				TAB + "A;"
+				);
+		// @formatter:on
+	}
+
+	@Test
+	public void tokenCardinality() {
+		// @formatter:off
+		assertFormattedGrammar(
+				"Real hidden(): INT ? '.' (EXT_INT | INT);"
+				,
+				"Real hidden():" + NL
+				+ TAB + "INT? '.' (EXT_INT | INT);"
+				);
+		// @formatter:on
+	}
+
+	@Test
+	public void unorderedGroups() {
+		// @formatter:off
+		assertFormattedGrammar(
+				"CopyFieldNameToVariableStmt:" + NL +
+				"  'FIELD-NAME-TO-VARIABLE' ((',' 'SCREEN' '=' '(' line=INT ',' column=INT ')') &" + NL +
+				TAB + TAB + TAB + "(','     'VAR' '=' name=ID) &" + NL +
+				TAB + TAB + TAB + "(',' 'TYPE' '='    'REPLACE')?);"
+				,
+				"CopyFieldNameToVariableStmt:" + NL +
+				TAB + "'FIELD-NAME-TO-VARIABLE' ((',' 'SCREEN' '=' '(' line=INT ',' column=INT ')') &" + NL +
+				TAB + "(',' 'VAR' '=' name=ID) &" + NL +
+				TAB + "(',' 'TYPE' '=' 'REPLACE')?);"
+				);
+		// @formatter:on
+	}
+
+	@Test
+	public void guardExpressions() {
+		// @formatter:off
+		assertFormattedGrammar(
+				"AssignmentExpression <In, Yield> returns Expression:" + NL +
+				TAB + "<Yield> YieldExpression<In>" + NL +
+				TAB + "| RelationalExpression<In,Yield> (=> ({AssignmentExpression.lhs=current} op='=') rhs=AssignmentExpression<In,Yield>)?" + NL +
+				";"
+				,
+				"AssignmentExpression <In, Yield> returns Expression:" + NL +
+				TAB + "<Yield> YieldExpression<In>" + NL +
+				TAB + "| RelationalExpression<In, Yield> (=> ({AssignmentExpression.lhs=current} op='=')" + NL +
```

Comment:
It would be really great to get rid of the newline in this example before the last `rhs=AssignmentExpression<In, Yield>`